### PR TITLE
First draft of the client gem MVP

### PR DESCRIPTION
**Pull request by lostie**
The intent is to provide calls to basic functionality of the existent supported calls to the API.

Adds a main API module where we can:
- check the version of the gem
- configure the client options
- retrieve an instance of the client itself

The client currently supports calls to the applications endpoints to create and fetch.

**NOTE:** Currently the gem has an its own version of the JSON schemas for the API endpoints to validate that the returned payloads from the server is the one expected (set by the _contract_). Ideally these schemas should be move to a common centralised location so that the server can update those when changes are made to the contracts and the client can fetch the most updated version of them and assert the current contract is still valid or requires changes.

**NOTE 2:** I haven't yet done any work to setup the project to publish the gem, but I believe if we're happy with the changes we can add that extra bit fairly easily as long we had permissions to push to the rubygems.org MOJ/LAA account.

### CODE REVIEW

File: laa-apply-for-legalaid-api-client.gemspec [hunk 1]
```diff
@@ -0,0 +1,32 @@
+lib = File.expand_path('../lib', __FILE__)
+$LOAD_PATH.unshift(lib) unless $LOAD_PATH.include?(lib)
+require 'laa/apply_for_legal_aid/api/version'
+
+Gem::Specification.new do |spec|
+  spec.name          = 'laa-apply-for-legalaid-api-client'
+  spec.version       = Laa::ApplyForLegalAid::Api::VERSION
+  spec.authors       = ['Sergio Marques', 'Ministry of Justice']
+  spec.email         = ['sergio.marques@digital.justice.gov.uk']
```
Comment (jsugarman) [inline laa-apply-for-legalaid-api-client.gemspec]: would add the 'tools@digital.justice.gov.uk' here, and/or and laa digital equivalent account.

File: lib/laa/apply_for_legal_aid/api/client.rb [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+          Request.new(method_name, path, connection).call(req_options)
+        end
+
+        private
+
+        def connection
+          @connection ||= initialize_connection
+        end
+
+        def reset_connection
```
Comment (ddippolito) [inline lib/laa/apply_for_legal_aid/api/client.rb]: Is this ever being used?

File: spec/laa/apply_for_legal_aid/api/client_spec.rb [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  let(:request_instance) { instance_double(Laa::ApplyForLegalAid::Api::Request) }
+  let(:request_params) { ['arg1', 'arg2', request_options] }
+  let(:request_options) { { data: { foo: 'bar' }} }
+  let(:mocked_response) { double(:http_response) }
+
+  subject(:client) { described_class.new(host, options) }
+
+  %w[get post put patch delete head].each do |http_method|
+    it "responds to a #{http_method} request" do
+      expect(Laa::ApplyForLegalAid::Api::Request)
```
Comment (ddippolito) [inline spec/laa/apply_for_legal_aid/api/client_spec.rb]: I am not sure about this one.

I would have moved this outside the expectation in a `before` block considering it part of my setup.

File: spec/laa/apply_for_legal_aid/api/client_spec.rb [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+        .to receive(:new)
+        .with(http_method.to_sym, '/some-resource', an_instance_of(Faraday::Connection))
+        .and_return(request_instance)
+      expect(request_instance).to receive(:call).with(request_options).and_return(mocked_response)
```
Comment (ddippolito) [inline spec/laa/apply_for_legal_aid/api/client_spec.rb]: I would have probably moved this in a different `it` block specifically for responses

File: lib/laa/apply_for_legal_aid/api/client.rb (hunk 2 above)
Comment (lostie) [reply lib/laa/apply_for_legal_aid/api/client.rb]: Not really, I had an idea to use it, but for this MVP I ended up not doing it.

File: spec/laa/apply_for_legal_aid/api/client_spec.rb (hunk 3 above)
Comment (lostie) [reply spec/laa/apply_for_legal_aid/api/client_spec.rb]: It's both setup and expectation, hence defining here to enforce is being called.

Arguably these sort of tests tend to be kind of dumb, but for unit test purposes, I avoided to test the whole end to end, so I've decided to go with this approach.

**Comment by lostie** [comment]
I've addressed the two main things with `reset_connection` not being used and the the extra email in the gem spec.

If there's any further improvements you would like me to do, please create an issue and I'll try to address them

**Pull request merged**